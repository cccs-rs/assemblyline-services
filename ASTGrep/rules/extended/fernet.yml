# yaml-language-server: $schema=https://raw.githubusercontent.com/ast-grep/ast-grep/main/schemas/rule.json

id: obfuscation-fernet
message: Executing code dynamically $CODE
severity: error # error, warning, info, hint
language: Python
rule:
  pattern: Fernet($PASS).decrypta($DATA)
metadata:
  some: metadata
  extended-obfuscation: yes
note: test

# utils: Extract repeated rule as local utility here.
# note: Add detailed explanation for the rule.