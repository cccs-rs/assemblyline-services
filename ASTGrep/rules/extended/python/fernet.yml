# yaml-language-server: $schema=https://raw.githubusercontent.com/ast-grep/ast-grep/main/schemas/rule.json

id: obfuscation-python-fernet
message: Contains code obfuscated with Fernet encryption
severity: error # error, warning, info, hint
language: Python
rule:
  any:
    - pattern:
        context: Fernet("$FERNET_KEY").decrypt("$FERNET_SOURCE")
        strictness: signature
    - pattern:
        context: $X.decrypt("$FERNET_SOURCE")
        selector: expression_statement
        strictness: signature
      follows:
        stopBy: end
        pattern:
          context: $X = Fernet("$FERNET_KEY")
          selector: expression_statement
          strictness: signature
    # TODO: other variations
metadata:
  extended-obfuscation: yes
  deobfuscate: |
    {
      "type": "extract",
      "steps": [
        {"func": "encode", "source": "FERNET_KEY", "encoding": "utf-8"},
        {"func": "encode", "source": "FERNET_SOURCE", "encoding": "utf-8"},
        {"func": "fernet"}
      ]
    }
