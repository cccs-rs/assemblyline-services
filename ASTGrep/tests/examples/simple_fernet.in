from fernet import Fernet

data = Fernet(b'1BncuowKwxoqx1edWbvkNRzLyq2LrJPHkO8dV_ZvGOc=').decrypt(b'gAAAAABmunksO4itD8DCrWMKKNhSglyuXQpQVXLKpg5JlmYbu6KXxjLzWR0Z3DqCtupx9zaztagzi2ujJtD6wBBPeqEfXER7pvuR2wY-OWKxmz4lkN7D5is=')
exec(data)