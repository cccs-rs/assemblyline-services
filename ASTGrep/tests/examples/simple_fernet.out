from fernet import Fernet

data = "print('hello world')"
print('hello world')